### PR TITLE
Spring Boot / Spring Boot Testing /  Add tests for a Book functionality

### PR DESCRIPTION
**Pull request by IvanovYurii**
1.	Changes:
    •	The book creation method (**create**) has been rewritten. Added category existence check.
    •	The book data update method (**updateById**) has been rewritten. An update request can have empty field values. Mandatory entry fields will remain unchanged, optional fields will be empty.
    •	The **getBooksByCategoryId** method has been moved from **CategoryService** to **BookService**.
    •	Import lombok.experimental.**Accessors** for entities that supported it.
    •	**BookMapper**. Added a null check to the setCategories method.
    •	Added necessary data to a**pplication.properties, pom.xml, docker-compose.yml** files.
    •	In **CategoryServiceImpl**, the logic that was repeated was moved to a separate **assertCategoryExistsById** method
    •	A default test is returned in **SpringBootIntroApplicationTests**.
2.	Fixed:
    •	Fixed error return texts in **CustomGlobalExceptionHandler**.
    •	**Category** entity field **category_description was accidentally marked as not null**. Fixed in all related files: **Category, CreateCategoryRequestDto, 10-create-categories-table.yaml**
3.	Added:
•	**UpdateBookRequestDto** for the ability to send an update request with empty fields.
•	Service level tests: **BookServiceImplTest, CategoryServiceImplTest.**
•	application.properties in the tests is configured to use the **docker container**.
•	The database folder has been added, which contains files that allow you to work with the database (create, destroy, modify tables and data) with the help of sql queries.
•	Added integration tests **BookStoreControllerIntegrationTest**, **CategoryControllerIntegrationTest**.

4. **Screenshot with the test coverage result**

Screenshot with the test coverage result

### CODE REVIEW

File: src/main/java/ivanov/springbootintro/mapper/BookMapper.java [hunk 1]
```diff
@@ -32,12 +33,16 @@ default void setCategoryIds(@MappingTarget BookDto bookDto, Book book) {
...
     @AfterMapping
     default void setCategories(@MappingTarget Book book, CreateBookRequestDto requestDto) {
-        book.setCategories(requestDto.categoryIds().stream()
-                .map(categoryId -> {
-                    Category category = new Category();
-                    category.setId(categoryId);
-                    return category;
-                })
-                .collect(Collectors.toSet()));
+        if (requestDto.categoryIds() != null) {
```
Comment (Ivan95kos) [inline src/main/java/ivanov/springbootintro/mapper/BookMapper.java]: You can use the Objects.nonNull()
```suggestion
        if (Objects.nonNull(requestDto.categoryIds())) {
```

File: src/main/java/ivanov/springbootintro/service/impl/BookServiceImpl.java [hunk 2]
```diff
@@ -24,14 +31,22 @@ public class BookServiceImpl implements BookService {
...
     public BookDto create(CreateBookRequestDto requestDto) {
         if (bookRepository.findByIsbn(requestDto.isbn()).isPresent()) {
-            throw new EntityAlreadyPresentException("Book with Isbn " + requestDto.isbn()
-                    + " is all ready present");
+            throw new EntityAlreadyPresentException("Can't create a book. Book with Isbn "
+                    + requestDto.isbn() + " is already present");
         }
-        Book book = bookMapper.toEntity(requestDto);
+        Book book = new Book();
+        if (requestDto.categoryIds() != null && !requestDto.categoryIds().isEmpty()) {
```
Comment (Ivan95kos) [inline src/main/java/ivanov/springbootintro/service/impl/BookServiceImpl.java]: You can use the Objects.nonNull()

File: src/main/java/ivanov/springbootintro/service/impl/BookServiceImpl.java [hunk 3]
```diff
@@ -44,42 +59,45 @@ public List<BookDto> findAll(Pageable pageable) {
...
+            validateIsbnUniqueness(requestDto.isbn(), id);
+            book.setIsbn(requestDto.isbn());
+        } else {
+            book.setIsbn(book.getIsbn());
         }
-        return bookMapper.toDto(bookRepository.save(book));
+        book.setTitle(requestDto.title() != null && !requestDto.title().isEmpty()
+                ? requestDto.title() : book.getTitle());
+        book.setTitle(requestDto.title() != null && !requestDto.title().isEmpty()
+                ? requestDto.title() : book.getTitle());
```
Comment (Ivan95kos) [inline src/main/java/ivanov/springbootintro/service/impl/BookServiceImpl.java]: You do the same twice
```suggestion
        book.setTitle(Objects.nonNull(requestDto.title()) && !requestDto.title().isEmpty()
                ? requestDto.title() : book.getTitle());
```

File: src/main/java/ivanov/springbootintro/service/impl/BookServiceImpl.java [hunk 4]
```diff
@@ -44,42 +59,45 @@ public List<BookDto> findAll(Pageable pageable) {
...
+    public BookDto updateById(UpdateBookRequestDto requestDto, Long id) {
+        Book book = assertBookExistsById(id);
+        if (requestDto.isbn() != null && !requestDto.isbn().isEmpty()) {
+            validateIsbnUniqueness(requestDto.isbn(), id);
+            book.setIsbn(requestDto.isbn());
+        } else {
+            book.setIsbn(book.getIsbn());
         }
-        return bookMapper.toDto(bookRepository.save(book));
+        book.setTitle(requestDto.title() != null && !requestDto.title().isEmpty()
```
Comment (Ivan95kos) [inline src/main/java/ivanov/springbootintro/service/impl/BookServiceImpl.java:76]: It is better to move to the mapper, than to check every field here

**Review by Ivan95kos — CHANGES_REQUESTED** [review]
Please, fix all checks

**Review by kozhukhovsky — APPROVED** [review]

**Pull request merged**